### PR TITLE
[new release] http, cohttp, cohttp-top, cohttp-server-lwt-unix, cohttp-mirage, cohttp-lwt, cohttp-lwt-unix, cohttp-lwt-jsoo, cohttp-eio, cohttp-curl, cohttp-curl-lwt, cohttp-curl-async and cohttp-async (6.0.0~beta 1) 

### PR DESCRIPTION
**Pull request by mseri**
CHANGES:

- move generic client and server signatures to cohttp and use them across all packges (mefyl https://github.com/mirage/ocaml-cohttp/pull/984)
- cohttp-eio: Complete rewrite to follow common interfaces and behaviors. (mefyl https://github.com/mirage/ocaml-cohttp/pull/984)
- cohttp-eio: Add Client.make_generic and HTTPS support. (talex5 https://github.com/mirage/ocaml-cohttp/pull/1002)

### CODE REVIEW

**Comment by hannesm** [comment]
from the changelog, there's only "cohttp-eio" being changed -- why not publish only that package then?

**Comment by mseri** [comment]
It is a very good point. Mostly because of the locked version constraints.

I can double check that is only eio and modify the pr and the version locks to reflect this (it will be weird that cohttp-eio beta will use cohttp alpha but life is such I guess :P)

**Comment by hannesm** [comment]
> It is a very good point. Mostly because of the locked version constraints. I can double check that is only eio and modify the pr and the version locks to reflect this (it will be weird that cohttp-eio beta will use cohttp alpha but life is such I guess :P)

that's what I did in tls-eio 0.15.5 https://github.com/ocaml/opam-repository/pull/22366 and tls-eio 0.17.2 https://github.com/ocaml/opam-repository/pull/24515 ... looks weird, but it saves quite some computation on CI machines and opam clients (I understand this only partially fits this PR, as you mark it as avoid-version (which works for opam 2.1.0+ clients)).

**Comment by mseri** [comment]
I actually cannot do it. There are some changes in cohttp to make the client and server interfaces generic and suitable for all packages. The change should be backward compatible though. I edited the message and will add a mention also in cohttp's CHANGELOG

**Comment by mseri** [comment]
In any case you make a very good point. I will be more careful with future PRs

**Comment by avsm** [comment]
A lot of 404s on the archives; did the rearrangement of the PR delete the beta1 tbzs?

**Comment by mseri** [comment]
Ah you are right. Updated

File: packages/cohttp-lwt/cohttp-lwt.6.0.0~beta1/opam [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+license: "ISC"
+homepage: "https://github.com/mirage/ocaml-cohttp"
+doc: "https://mirage.github.io/ocaml-cohttp/"
+bug-reports: "https://github.com/mirage/ocaml-cohttp/issues"
+depends: [
+  "dune" {>= "3.0"}
+  "ocaml" {>= "4.08"}
+  "http" {= version}
+  "cohttp" {= version}
+  "lwt" {>= "2.5.0"}
```
Comment (mseri) [inline packages/cohttp-lwt/cohttp-lwt.6.0.0~beta1/opam:31]:
```suggestion
  "lwt" {>= "5.4.0"}
```
Uses Lwt.dont_wait

**Comment by mseri** [comment]
There seems to be a regression
```

#=== ERROR while compiling prometheus-app.1.2 =================================#
# context              2.2.0~alpha3~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.14/.opam-switch/build/prometheus-app.1.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p prometheus-app -j 127
# exit-code            1
# env-file             ~/.opam/log/prometheus-app-7-51de6d.env
# output-file          ~/.opam/log/prometheus-app-7-51de6d.out
### output ###
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlc.opt -w -40 -g -bin-annot -I app/.prometheus_app.objs/byte -I /home/opam/.opam/4.14/lib/angstrom -I /home/opam/.opam/4.14/lib/asetmap -I /home/opam/.opam/4.14/lib/astring -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigstringaf -I /home/opam/.opam/4.14/lib/cohttp -I /home/opam/.opam/4.14/lib/cohttp-lwt -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/http -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/lwt -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/prometheus -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stringext -I /home/opam/.opam/4.14/lib/uri -I /home/opam/.opam/4.14/lib/uri-sexp -intf-suffix .ml -no-alias-deps -o app/.prometheus_app.objs/byte/prometheus_app.cmo -c -impl app/prometheus_app.ml)
# File "app/prometheus_app.ml", line 157, characters 6-27:
# 157 |       Server.respond_string ~status:`OK ~headers ~body ()
#             ^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value Server.respond_string
# (cd _build/default && /home/opam/.opam/4.14/bin/ocamlopt.opt -w -40 -g -I app/.prometheus_app.objs/byte -I app/.prometheus_app.objs/native -I /home/opam/.opam/4.14/lib/angstrom -I /home/opam/.opam/4.14/lib/asetmap -I /home/opam/.opam/4.14/lib/astring -I /home/opam/.opam/4.14/lib/base64 -I /home/opam/.opam/4.14/lib/bigstringaf -I /home/opam/.opam/4.14/lib/cohttp -I /home/opam/.opam/4.14/lib/cohttp-lwt -I /home/opam/.opam/4.14/lib/fmt -I /home/opam/.opam/4.14/lib/http -I /home/opam/.opam/4.14/lib/http/__private__/http_bytebuffer -I /home/opam/.opam/4.14/lib/logs -I /home/opam/.opam/4.14/lib/lwt -I /home/opam/.opam/4.14/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.14/lib/prometheus -I /home/opam/.opam/4.14/lib/re -I /home/opam/.opam/4.14/lib/seq -I /home/opam/.opam/4.14/lib/sexplib0 -I /home/opam/.opam/4.14/lib/stringext -I /home/opam/.opam/4.14/lib/uri -I /home/opam/.opam/4.14/lib/uri-sexp -intf-suffix .ml -no-alias-deps -o app/.prometheus_app.objs/native/prometheus_app.cmx -c -impl app/prometheus_app.ml)
# File "app/prometheus_app.ml", line 157, characters 6-27:
# 157 |       Server.respond_string ~status:`OK ~headers ~body ()
#             ^^^^^^^^^^^^^^^^^^^^^
# Error: Unbound value Server.respond_string
```
Please don't merge

**Comment by mseri** [comment]
@samoht do you understand why irmin is failing now? The signature of IO should not have changed in cohttp

**Comment by samoht** [comment]
@mseri 

A diff between `alpha2` and `beta1` shows this:
```diff
 module type Server = sig
-  module IO : IO
```

This explains why this is breaking `irmin-graphql`

**Comment by mseri** [comment]
Seems like it is useful to have it there, what do you think? Should I try to have it there or just add an upper bound to irmin?

**Comment by samoht** [comment]
I think it'd be best not to break that API if possible - enough changes in the rest already :-)

**Comment by mseri** [comment]
Closing in favour of the upcoming beta 2

**Pull request closed without merging**